Add 'outdated' Command

l0nax opened the pull request:
The `outdated` Command shows the User if there are Releases that are outdated.

The normal Output is:
```
NAME              	NAMESPACE 	INSTALLED VERSION	LATEST VERSION	CHART              
kube-state-metrics	monitoring	2.3.1            	2.4.1         	kube-state-metrics 
prometheus        	monitoring	6.18.0           	8.2.2         	prometheus-operator
```

Help:
```
This Command lists all releases which are outdated.

By default, the output is printed in a Table but you can change this behavior
with the '--output' Flag.

Usage:
  helm outdated [flags]

Aliases:
  outdated, od

Flags:
  -a, --all              show all releases, not just the ones marked deployed or failed
  -A, --all-namespaces   list releases across all namespaces
  -d, --date             sort by release date
      --deployed         show deployed releases. If no other is specified, this will be automatically enabled
      --devel            use development versions (alpha, beta, and release candidate releases), too. Equivalent to version '>0.0.0-0'.
      --failed           show failed releases
  -h, --help             help for outdated
  -m, --max int          maximum number of releases to fetch (default 256)
      --offset int       next release name in the list, used to offset from start value
  -o, --output format    prints the output in the specified format. Allowed values: table, json, yaml (default table)
      --pending          show pending releases
  -r, --reverse          reverse the sort order
  -q, --short            output short (quiet) listing format
      --superseded       show superseded releases
      --uninstalled      show uninstalled releases
      --uninstalling     show releases that are currently being uninstalled
```

The Command support also checking for _pre-release_ Chart Versions and searching in all Namespaces.


Closes #7014

Commit:
Add 'outdated' Command Skeleton

Signed-off-by: Emanuel Bennici <eb@fabmation.de>

Commit:
Implement outdated Cmd Functionality

Signed-off-by: Emanuel Bennici <eb@fabmation.de>

Commit:
Add outdated Cmd to root.go

Signed-off-by: Emanuel Bennici <eb@fabmation.de>

Commit:
Correct Error messages in outdated Cmd

Signed-off-by: Emanuel Bennici <eb@fabmation.de>

Commit:
Add Repository Settings to Search Options

Signed-off-by: Emanuel Bennici <eb@fabmation.de>

Commit:
Fix 'chart not found' Bug in outdated Command

Signed-off-by: Emanuel Bennici <eb@fabmation.de>

Commit:
Fix GolangCI Style Error

Signed-off-by: Emanuel Bennici <eb@fabmation.de>

Commit:
Update 'devel' flag usage in outdated Cmd

Signed-off-by: Emanuel Bennici <eb@fabmation.de>

vdboor commented:
Thanks for this good first step!
I wonder, where does the "existing < available" check happen? I can't find that in this code

l0nax commented:
Locally it compares the Versions.
I will push the latest changes when I'm in the Office.

I forgot to push my latest changes 😅

Commit:
Add semver checks to outdated Command

Signed-off-by: Emanuel Bennici <eb@fabmation.de>

Commit:
Add new Flags to outdated Command

All the flags supported by the 'List' action (except of the 'filter'
flag) are added to the outdated Command.

Signed-off-by: Emanuel Bennici <eb@fabmation.de>

Commit:
Fix CircleCI 'test-style' error

The Code is correct and works correctly, but CircleCI claims that it is
wrong.

Signed-off-by: Emanuel Bennici <eb@fabmation.de>

l0nax commented:
This PR is ready to be reviewed

l0nax commented:
For everyone who wants to use this Command, I updated (_temporary, because I'm hoping that this Pull-Request will be accepted_) my [`whatup`](https://github.com/fabmation-gmbh/helm-whatup/releases/tag/v0.3.0) Fork and version _0.3.0_ now supports helm v3.

hickeyma reviewed (commented):
@bacongobbler @l0nax From reading the comments in #7014, I am unsure if this was functionality that was agreed upon to be pulled into core. Can you confirm this?

l0nax commented:
@hickeyma I do not understand your question. Can you please clarify it?

hickeyma commented:
@l0nax From reading though the comments in #7014, I cannot see where it is mentioned that we want the `outdated` command in Helm core? 

@bacongobbler What is your opinion on this?

bacongobbler commented:
As it currently stands, I have concerns bringing in this feature into core. Currently, there's no [downloader plugin support](https://helm.sh/docs/topics/plugins/#downloader-plugins), no test coverage and no documentation available.

If those issues are tackled, then I wouldn't have any concerns bringing it into core, assuming @l0nax is willing to continue working on it should future feature requests come up.

hickeyma commented:
Ok, that seems to me like there needs to be a proposal first detailing the feature with all the considerations taken into account to make this a rounded/fully supported feature?

@l0nax Would you be interested in doing this proposal?

l0nax commented:
> @l0nax Would you be interested in doing this proposal?
I haven't written any proposal before, but I can try it. So, yes I would be interested in creating the proposal.

> Currently, there's no downloader plugin support, no OCI support [...]

This command doesn't download anything. It uses the helm repository Index "DB" to search for outdated Charts. So I do not understand what do you mean with this.
And about the _OCI support_ stuff, with `helm repo update`, helm will create the index of all Repositories – and its charts – so this Command does not need to "support" OCI repositories because helm does it anyway/ under the hood.
Please correct me if I'm wrong @bacongobbler

l0nax commented:
And, yes, I would maintain it if any feature request or any other stuff comes up.

ms32035 commented on cmd/helm/outdated.go:
```diff
@@ -0,0 +1,231 @@
+	for _, r := range releases {
+		// search if it exists a newer Chart in the Chart-Repository
+		repoResult, err := searchChart(results, r.Chart.Name(), r.Chart.Metadata.Version, devel)
+		if err != nil {
```
This cause the whole plugin to error if there's just one chart that was deployed from local `Could not find any Repo which contains ...`. Even if that's intentional, there should be a parameter to turn such a case into a warning.

l0nax replied:
Done with commit 50f1deb1ed00e64dfa9aea964927a065a4a86fab

Commit:
Add flag to ignore charts where no repo was found

This new flag ('--ignore-repo') provides the functionality to allow the
user to continue using the 'outdated' Command even when no Repo was
found which holds the specific Chart.

Signed-off-by: Emanuel Bennici <eb@fabmation.de>

l0nax commented:
> @l0nax Would you be interested in doing this proposal?
@hickeyma Yes, I would create such I proposal. Is there any starting point on how to write such a proposal?

hickeyma commented:
@l0nax Take look at any issues with label `proposal` as a starting point e.g. #7260

l0nax commented:
Ok thanks @hickeyma 

Should this be a new Issue (marked as `//WIP: `) or should the current Issue (#7014) be updated?

hickeyma commented:
@l0nax Might be worth opening a new issue and reference that issue.

l0nax commented:
Ok

bacongobbler commented:
closing as this PR has been inactive for 7 months.
